internal/integration/bounded_harvester.go

```diff
@@ -0,0 +1,166 @@
+	}
+
+	return h
+}
+
+// BoundedHarvesterCfg stores the configurable values for boundedHarvester
+type BoundedHarvesterCfg struct {
+	// MetricCap is the number of metrics to store in memory before triggering a HarvestNow action regardless of
+	// HarvestPeriod. It will directly influence the amount of memory that nri-prometheus allocates.
+	// A value of 10000 is rougly equivalent to 500M in RAM in the tested scenarios
+	MetricCap int
+
+	// HarvestPeriod specifies the period that will trigger a HarvestNow action for the inner harvester.
+	// It is not necessary to decrease this value further, as other conditions (namely the MetricCap) will also trigger
+	// a harvest action.
+	HarvestPeriod time.Duration
+
+	// MinReportInterval Specifies the minimum amount of time to wait before reports.
+	// This will be always enforced, regardless of HarvestPeriod and MetricCap.
+	MinReportInterval time.Duration
+
+	// DisablePeriodicReporting prevents bindHarvester from spawning the periodic report routine.
+	// It also causes an already spawned reporting routine to be stopped on the next interval.
+	DisablePeriodicReporting bool
+}
+
+// BoundedHarvesterDefaultHarvestPeriod is the default harvest period. Since harvests are also triggered by stacking
+// metrics, there is no need for this to be very low
+const BoundedHarvesterDefaultHarvestPeriod = 5 * time.Second
+
+// BoundedHarvesterDefaultMetricsCap is the default number of metrics stack before triggering a harvest. 10000 metrics
+// require around 500MiB in our testing setup
+const BoundedHarvesterDefaultMetricsCap = 10000
+
+// BoundedHarvesterDefaultMinReportInterval is the default and minimum enforced harvest interval time. No harvests will
+// be issued if previous harvest was less than this value ago (except for those triggered with HarvestNow)
+const BoundedHarvesterDefaultMinReportInterval = 200 * time.Millisecond
+
+// boundedHarvester is a harvester implementation and wrapper that keeps count of the number of metrics that are waiting
+// to be harvested. Every small period of time (BoundedHarvesterCfg.MinReportInterval), if the number of accumulated
+// metrics is above a given threshold (BoundedHarvesterCfg.MetricCap), a harvest is triggered.
+// A harvest is also triggered in periodic time intervals (BoundedHarvesterCfg.HarvestPeriod)
+// boundedHarvester will never trigger harvests more often than specified in BoundedHarvesterCfg.MinReportInterval.
+type boundedHarvester struct {
+	BoundedHarvesterCfg
+
+	mtx sync.Mutex
+
+	storedMetrics int
+	lastReport    time.Time
+
+	stopper chan struct{}
+	stopped bool
+
+	inner harvester
+}
+
+// RecordMetric records the metric in the underlying harvester and reports all of them if needed
+func (h *boundedHarvester) RecordMetric(m telemetry.Metric) {
+	h.mtx.Lock()
+	h.storedMetrics++
+	h.mtx.Unlock()
+
+	h.inner.RecordMetric(m)
+}
+
+// HarvestNow forces a new report
+func (h *boundedHarvester) HarvestNow(ctx context.Context) {
```

Comment:
do we need this function to "satisfy" an interface requirement?
Otherwise if it is not needed by the code I would get rid of it and of the force parameter

Reply:
It is mainly used for tests, but also to keep compatibility with other `Harvester` implementations. It would be impossible to wrap/compound `Harvester`s without it, in fact, `boundedHarvester` uses it precisely for this purpose :)